File: rotateArray.py
# -*- coding:utf-8 -*-
#³¬Ê±
class Solution:
    def minNumberInRotateArray(self, rotateArray):
        # write code here
        self.rotateArray=rotateArray
        self.lens=len(self.rotateArray)
        self.index1=0
        self.index2=self.lens-1
        self.indexmid=0
        if self.lens==0:
            return None
        while self.rotateArray[self.index1]>=self.rotateArray[self.index2]:
            if self.index2-self.index1==1:
                self.indexmid=self.index2
                break
            self.indexmid=(self.index1+self.index2)/2
            if (self.rotateArray[self.index1]==self.rotateArray[self.index2])and(self.rotateArray[self.index1]==self.rotateArray[self.indexmid]):
                return self.shunxu()
            if self.rotateArray[self.indexmid]>=self.rotateArray[self.index1]:
                self.rotateArray[self.index1]=self.rotateArray[self.indexmid]
            else:
                self.rotateArray[self.index2]=self.rotateArray[self.indexmid]
        return self.rotateArray[self.indexmid]
    def shunxu(self):
        self.result=self.rotateArray[0]
        for i in range(self.rotateArray[self.index1+1],self.rotateArray[self.index2]+1):
            if self.rotateArray[i]<self.result:
                self.result=self.rotateArray[i]
        return self.result

File: rotateArray.cpp
class Solution {
public:
    int minNumberInRotateArray(vector<int> rotateArray) {
        int length=rotateArray.size();
        if(length<=0){
            throw new exception();
        }
        int index1=0,index2=length-1;
        int indexmid=index1;
        while(rotateArray[index1]>=rotateArray[index2]){
            if(index2-index1==1){
                indexmid=index2;
                break;
            }
            indexmid=(index1+index2)/2;
            if(rotateArray[index1]==rotateArray[index2]&&rotateArray[index1]==rotateArray[indexmid]){
                return shunxu(rotateArray,index1,index2);
            }
            if(rotateArray[indexmid]>=rotateArray[index1]){
                index1=indexmid;
            }
            else{
                index2=indexmid;
            }
        }
        return rotateArray[indexmid];
    }
    int shunxu(vector<int> rotateArray,int index1,int index2){
        int a=rotateArray[index1];
        for(int i=index1+1;i<=index2;i++){
            if(a>rotateArray[i]){
                a=rotateArray[i];
            }
        }
        return a;
    }
};